ESQL: Block access benchmark

alex-spies opened the pull request:
Benchmark how long it takes to access elements of `IntBlock`s, `BooleanBlock`s, etc.

This helps guard against performance regressions when refactoring Block and Vector classes.

alex-spies committed:
Draft Block access benchmark

nik9000 commented on benchmarks/src/main/java/org/elasticsearch/benchmark/compute/operator/BlockBenchmark.java:
```diff
@@ -0,0 +1,174 @@
+    }
+
+    @Benchmark
+    @OperationsPerInvocation(BlockBenchmark.NUM_BLOCKS_PER_RUN)
```
Kind of a shame that the OperationsPerInvocation can't be some multiple of the block size. To really read the time things take you have to divide by the number of entries in the block.

I wonder if we're better off using a fixed number of entries in the block so we can get that? Not sure.

alex-spies replied:
Hm, maybe the block length turns out to not be important - then I'll just set it to a fixed number!

alex-spies replied:
Set it to a constant number and made it so we take the time per position instead of per block.

nik9000 reviewed (commented):
Seems like the right direction, yeah!

alex-spies commented on benchmarks/src/main/java/org/elasticsearch/benchmark/compute/operator/BlockBenchmark.java:
```diff
@@ -0,0 +1,174 @@
+    @Param({ "array", "vector" })
+    public String blockKind;
+
+    @Setup(Level.Trial)
```
We can, this is equivalent.

alex-spies committed:
Use hard checksum check; read blocks idiomatically

alex-spies committed:
Use default setup annotation

alex-spies committed:
Merge remote-tracking branch 'upstream/main' into esql-block-access-benchmark

alex-spies committed:
Check checksums after iterations

alex-spies committed:
Use ByteSizeValue

alex-spies committed:
Remove leftovers

alex-spies committed:
Ensure all classes are loaded before first bench

Perform a small benchmark run at class loading time to ensure all
classes used in the benchmark are loaded before the first iteration.
This is important to avoid accidentally measuring some JVM optimizations
that will not be possible in production (method inlining in case of
small number of interface implementers or similar).

alex-spies committed:
Add random access benchmark

alex-spies committed:
Remove comment

alex-spies committed:
Remove obsolete access to position count

alex-spies committed:
Merge remote-tracking branch 'upstream/main' into esql-block-access-benchmark

alex-spies committed:
Benchmark also MVs and nulls

alex-spies committed:
Add boolean blocks

alex-spies committed:
Add boolean to actual parameter list

alex-spies committed:
Ignore nulls

alex-spies committed:
Add double data type

alex-spies committed:
Merge remote-tracking branch 'upstream/main' into esql-block-access-benchmark

alex-spies committed:
Add BytesRef type

alex-spies committed:
Choose fixed block length

alex-spies committed:
Add BigArrayBlocks

alex-spies committed:
Add BigArrayVectors

alex-spies committed:
Add constant vector blocks

alex-spies committed:
Remove leftovers

elasticsearchmachine commented:
Pinging @elastic/es-ql (Team:QL)

elasticsearchmachine commented:
Pinging @elastic/elasticsearch-esql (:Query Languages/ES|QL)

alex-spies committed:
Update comment

alex-spies commented:
Ran the bench on my machine to get a first idea:
```
Result "org.elasticsearch.benchmark.compute.operator.BlockBenchmark.run":
  2.161 ±(99.9%) 0.501 ns/op [Average]
  (min, avg, max) = (1.981, 2.161, 2.403), stdev = 0.222
  CI (99.9%): [1.660, 2.662] (assumes normal distribution)


# Run complete. Total time: 01:52:55

REMEMBER: The numbers below are just data. To gain reusable insights, you need to follow up on
why the numbers are the way they are. Use profilers (see -prof, -lprof), design factorial
experiments, perform baseline and negative tests that provide experimental control, make sure
the benchmarking environment is safe on JVM/OS/HW level, ask for reviews from the domain experts.
Do not assume the numbers tell you what you want them to tell.

Benchmark           (accessType)             (dataTypeAndBlockKind)  Mode  Cnt   Score   Error  Units
BlockBenchmark.run    sequential                      boolean/array  avgt    7   1.432 ± 0.022  ns/op
BlockBenchmark.run    sequential      boolean/array-multivalue-null  avgt    7   0.841 ± 0.004  ns/op
BlockBenchmark.run    sequential  boolean/big-array-multivalue-null  avgt    7   1.711 ± 0.107  ns/op
BlockBenchmark.run    sequential                     boolean/vector  avgt    7   2.292 ± 0.408  ns/op
BlockBenchmark.run    sequential           boolean/vector-big-array  avgt    7   3.261 ± 0.014  ns/op
BlockBenchmark.run    sequential               boolean/vector-const  avgt    7   1.983 ± 0.547  ns/op
BlockBenchmark.run    sequential                     BytesRef/array  avgt    7   8.471 ± 0.090  ns/op
BlockBenchmark.run    sequential     BytesRef/array-multivalue-null  avgt    7   6.654 ± 0.047  ns/op
BlockBenchmark.run    sequential                    BytesRef/vector  avgt    7   8.442 ± 0.074  ns/op
BlockBenchmark.run    sequential              BytesRef/vector-const  avgt    7   2.720 ± 0.010  ns/op
BlockBenchmark.run    sequential                       double/array  avgt    7   1.748 ± 0.015  ns/op
BlockBenchmark.run    sequential       double/array-multivalue-null  avgt    7   0.797 ± 0.005  ns/op
BlockBenchmark.run    sequential   double/big-array-multivalue-null  avgt    7   1.313 ± 0.005  ns/op
BlockBenchmark.run    sequential                      double/vector  avgt    7   2.298 ± 0.399  ns/op
BlockBenchmark.run    sequential            double/vector-big-array  avgt    7   3.100 ± 0.032  ns/op
BlockBenchmark.run    sequential                double/vector-const  avgt    7   2.141 ± 0.502  ns/op
BlockBenchmark.run    sequential                          int/array  avgt    7   1.319 ± 0.006  ns/op
BlockBenchmark.run    sequential          int/array-multivalue-null  avgt    7   0.753 ± 0.005  ns/op
BlockBenchmark.run    sequential      int/big-array-multivalue-null  avgt    7   1.315 ± 0.076  ns/op
BlockBenchmark.run    sequential                         int/vector  avgt    7   2.320 ± 0.522  ns/op
BlockBenchmark.run    sequential               int/vector-big-array  avgt    7   3.061 ± 0.012  ns/op
BlockBenchmark.run    sequential                   int/vector-const  avgt    7   2.237 ± 0.244  ns/op
BlockBenchmark.run    sequential                         long/array  avgt    7   1.343 ± 0.015  ns/op
BlockBenchmark.run    sequential         long/array-multivalue-null  avgt    7   0.831 ± 0.006  ns/op
BlockBenchmark.run    sequential     long/big-array-multivalue-null  avgt    7   1.394 ± 0.009  ns/op
BlockBenchmark.run    sequential                        long/vector  avgt    7   2.351 ± 0.459  ns/op
BlockBenchmark.run    sequential              long/vector-big-array  avgt    7   3.072 ± 0.015  ns/op
BlockBenchmark.run    sequential                  long/vector-const  avgt    7   2.156 ± 0.508  ns/op
BlockBenchmark.run        random                      boolean/array  avgt    7   1.490 ± 0.009  ns/op
BlockBenchmark.run        random      boolean/array-multivalue-null  avgt    7   0.779 ± 0.003  ns/op
BlockBenchmark.run        random  boolean/big-array-multivalue-null  avgt    7   1.300 ± 0.005  ns/op
BlockBenchmark.run        random                     boolean/vector  avgt    7   2.396 ± 0.300  ns/op
BlockBenchmark.run        random           boolean/vector-big-array  avgt    7   3.277 ± 0.013  ns/op
BlockBenchmark.run        random               boolean/vector-const  avgt    7   2.002 ± 0.533  ns/op
BlockBenchmark.run        random                     BytesRef/array  avgt    7  10.405 ± 0.087  ns/op
BlockBenchmark.run        random     BytesRef/array-multivalue-null  avgt    7   7.277 ± 0.070  ns/op
BlockBenchmark.run        random                    BytesRef/vector  avgt    7   9.598 ± 0.157  ns/op
BlockBenchmark.run        random              BytesRef/vector-const  avgt    7   2.403 ± 0.017  ns/op
BlockBenchmark.run        random                       double/array  avgt    7   2.013 ± 0.011  ns/op
BlockBenchmark.run        random       double/array-multivalue-null  avgt    7   1.085 ± 0.005  ns/op
BlockBenchmark.run        random   double/big-array-multivalue-null  avgt    7   1.703 ± 0.042  ns/op
BlockBenchmark.run        random                      double/vector  avgt    7   3.099 ± 0.017  ns/op
BlockBenchmark.run        random            double/vector-big-array  avgt    7   4.055 ± 0.028  ns/op
BlockBenchmark.run        random                double/vector-const  avgt    7   2.155 ± 0.517  ns/op
BlockBenchmark.run        random                          int/array  avgt    7   1.630 ± 0.011  ns/op
BlockBenchmark.run        random          int/array-multivalue-null  avgt    7   0.869 ± 0.017  ns/op
BlockBenchmark.run        random      int/big-array-multivalue-null  avgt    7   1.438 ± 0.009  ns/op
BlockBenchmark.run        random                         int/vector  avgt    7   2.572 ± 0.147  ns/op
BlockBenchmark.run        random               int/vector-big-array  avgt    7   3.434 ± 0.032  ns/op
BlockBenchmark.run        random                   int/vector-const  avgt    7   2.249 ± 0.244  ns/op
BlockBenchmark.run        random                         long/array  avgt    7   2.035 ± 0.015  ns/op
BlockBenchmark.run        random         long/array-multivalue-null  avgt    7   1.062 ± 0.008  ns/op
BlockBenchmark.run        random     long/big-array-multivalue-null  avgt    7   1.654 ± 0.059  ns/op
BlockBenchmark.run        random                        long/vector  avgt    7   3.046 ± 0.011  ns/op
BlockBenchmark.run        random              long/vector-big-array  avgt    7   4.083 ± 0.024  ns/op
BlockBenchmark.run        random                  long/vector-const  avgt    7   2.161 ± 0.501  ns/op
```
(Note: I didn't take particular precautions to guard against noise, didn't disable optimizations, didn't bind to single CPU, so we could also get more precise results.)

alex-spies committed:
Merge remote-tracking branch 'upstream/main' into esql-block-access-benchmark

ChrisHegarty reviewed (approved):
LGTM

nik9000 commented:
Looks useful! Let's land it. I imagine we'll play with this a lot over the next few years.
